### DISCUSSION TITLE
What's a good way to debug deep linking?

### DISCUSSION THREAD

**Discussion by tc** [csells/go_router#294] [General]
I have an app using go_router, I used `flutter build web` and uploaded to two s3 bucket/cloudfront cdn:
The only difference is an api server reference: `api-staging` vs `api`

Deep linking works here:
https://previewapp-staging.tempokit.com/workouts/5e8e48d30b7751766ae4ec15

Deep linking doesn't work and it redirects to the `/` route.
https://previewapp.tempokit.com/workouts/5e8e48d30b7751766ae4ec15

(i set both cloudfront cdns to use redirect errors to '/index.html')

```dart
    _router = GoRouter(
      refreshListenable: GoRouterRefreshStream(
        widget.authCubit.routeRefreshStream,
      ),
      redirect: (state) {
        final authState = widget.authCubit.state;

        if (authState is AuthStateLoaded) {
          currentToken = authState.token;
        }

        //handle redirect to main page  after user logs in or out
        if (oldToken != currentToken) {
          print('oldToken=$oldToken currentToken=$currentToken');

          oldToken = currentToken;
          var redirectRoute = '/app/home';
          // Prevent redirect loop
          if (redirectRoute != state.subloc) {
            print('new redirect ${redirectRoute}');
            return redirectRoute;
          }
        }
        return null;
      },
      debugLogDiagnostics: true,
      routes: [
        GoRoute(
          path: '/',
          redirect: (_) => '/app/home',
        ),
        GoRoute(
          path: '/workouts/:id',
          builder: (context, state) => WorkoutScreen(id: state.params['id']),
        ),
      ],
    );
  ```

What's a good way to debug? I set `debugLogDiagnostics: true` but see no useful info in the redirection or deep link failing.

I checked in the Firefox console and only see this as the difference:

```
Uncaught Error: Stack Overflow
    A https://previewapp.tempokit.com/main.dart.js:4061
```

In Chrome, I only see:

```
Refused to set unsafe header "User-Agent"
```

**Comment by csells** [comment]
have you [enabled debugging on the `GoRouter`](https://gorouter.dev/debugging)?

**Reply by tc** [reply]
i have `      debugLogDiagnostics: true,` 

i am building using `flutter build web --release`

I don't see output like 
```dart
GoRouter: known full paths for routes:
GoRouter:   => /
```
Is it forced off when in release mode?

I do see this output in local host when running via android studio, but the problem doesn't happen locally.

**Comment by csells** [comment]
The GR debug output only happens during debugging. You may want to provide a `NavigatorObserver` to see what's happening during your navigation during release mode.

**Reply by tc** [reply]
is there a way to force the GR debug logging in release mode?

It looks like i would be copy pasta [NavigatorObserver](https://github.com/csells/go_router/blob/main/go_router/example/lib/nav_observer.dart)  which looks very similar to the GR debug mode?

**Comment by csells** [comment]
Not currently. @johnpryan is that something that we can do with the logging API?

**Reply by tc** [reply]
Thanks for the tip I used the `MyNavObserver` in the https://github.com/csells/go_router/blob/main/go_router/example/lib/nav_observer.dart example and it helped guide me to my problem. (auth state redirection)

**Reply by johnpryan** [reply]
Good to hear!

**Comment by csells** [comment]
excellent
